src/resources/i2c_esp32.cc

```diff
@@ -18,274 +18,270 @@
-  i2c_port_t port_;
+  friend class I2cBusResource;
+  I2cBusResource* bus_;
+  i2c_master_dev_handle_t handle_;
 };
 
-MODULE_IMPLEMENTATION(i2c, MODULE_I2C);
+class I2cBusResource : public Resource, public DeviceList {
+ public:
+  TAG(I2cBusResource);
+  I2cBusResource(I2cResourceGroup* group, i2c_master_bus_handle_t handle)
+      : Resource(group)
+      , handle_(handle) {}
 
-PRIMITIVE(init) {
-  ARGS(int, frequency, int, sda, int, scl, bool, sda_pullup, bool, scl_pullup);
+  ~I2cBusResource() override;
 
-  i2c_port_t port = i2c_ports.any();
-  if (port == kInvalidPort) FAIL(ALREADY_IN_USE);
+  i2c_master_bus_handle_t handle() const { return handle_; }
 
-  ByteArray* proxy = process->object_heap()->allocate_proxy();
-  if (proxy == null) {
-    i2c_ports.put(port);
-    FAIL(ALLOCATION_FAILED);
-  }
+  void add_device(I2cDeviceResource* device);
+  void remove_device(I2cDeviceResource* device);
 
-  i2c_config_t conf;
-  memset(&conf, 0, sizeof(conf));
-  conf.mode = I2C_MODE_MASTER;
-  conf.sda_io_num = (gpio_num_t)sda;
-  conf.sda_pullup_en = sda_pullup ? GPIO_PULLUP_ENABLE : GPIO_PULLUP_DISABLE;
-  conf.scl_io_num = (gpio_num_t)scl;
-  conf.scl_pullup_en = scl_pullup ? GPIO_PULLUP_ENABLE : GPIO_PULLUP_DISABLE;
-  conf.master.clk_speed = frequency;
-  int result = i2c_param_config(port, &conf);
-  if (result != ESP_OK) FAIL(INVALID_ARGUMENT);
-  result = ESP_FAIL;
-  SystemEventSource::instance()->run([&]() -> void {
-    result = i2c_driver_install(port, I2C_MODE_MASTER, 0, 0, 0);
-#if defined(CONFIG_IDF_TARGET_ESP32C3) || defined(CONFIG_IDF_TARGET_ESP32C6) || defined(CONFIG_IDF_TARGET_ESP32S3)
-    i2c_set_timeout(port, (int)(log2(APB_CLK_FREQ / 1000.0 * kI2cTransactionTimeout)));
-#elif defined(CONFIG_IDF_TARGET_ESP32) || defined(CONFIG_IDF_TARGET_ESP32S2)
-    i2c_set_timeout(port, I2C_APB_CLK_FREQ / 1000 * kI2cTransactionTimeout);
-#else
-#error "Unsupported target"
-  // Go to the i2c section of the datasheet and check whether the value is used as
-  // a power or counts the individual cycles.
-  // For example:
-  //   https://www.espressif.com/sites/default/files/documentation/esp32-s3_technical_reference_manual_en.pdf#i2c
-#endif
-  });
-  if (result != ESP_OK) {
-    i2c_ports.put(port);
-    return Primitive::os_error(result, process);
+  I2cResourceGroup* resource_group() const {
+    return static_cast<I2cResourceGroup*>(Resource::resource_group());
   }
 
-  I2cResourceGroup* i2c = _new I2cResourceGroup(process, port);
-  if (!i2c) {
-    SystemEventSource::instance()->run([&]() -> void {
-      i2c_driver_delete(port);
-    });
-    i2c_ports.put(port);
-    FAIL(MALLOC_FAILED);
+ private:
+  i2c_master_bus_handle_t handle_;
+};
+
+I2cDeviceResource::~I2cDeviceResource() {
+  if (bus_ != null) {
+    bus_->remove_device(this);
   }
+}
 
-  proxy->set_external_address(i2c);
+I2cBusResource::~I2cBusResource() {
+  while (!DeviceList::is_empty()) {
+    // Removing the device doesn't delete the resource. That needs to be
```

Comment:
I think this is hard to understand. Both the bus and the devices are resources, so which one are you taking about here ("delete the resource")? Also, where is the separate deletion done?